### PR TITLE
#89 - add basic authentication option when the token is not provided …

### PR DESCRIPTION
**Pull request by tremes**
…in user config

### CODE REVIEW

**Comment by tremes** [comment]
I can write some test for this but I think I will need user credentials for the test clusters.

File: lib/config-loader.js [hunk 1]
```diff
@@ -35,7 +35,7 @@ function loadConfig (client) {
   return load(client.settings).then((config) => {
     client.apiUrl = `${config.cluster}/oapi/${config.apiVersion}`;
     client.kubeUrl = `${config.cluster}/api/${config.apiVersion}`;
-
+    client.authUrl = `${config.cluster}/oauth/authorize?response_type=token&client_id=openshift-challenging-client`;
```
[the remark this answers is not in the document]
Comment (tremes) [reply lib/config-loader.js]: I know only about https://docs.openshift.com/container-platform/3.10/architecture/additional_concepts/authentication.html and TBH I followed a pattern observed from oc calls.

**Comment by lholmquist** [comment]
@tremes could you update the readme with the use case for this

**Comment by tremes** [comment]
@lholmquist OK will do. I can also add some test but would need username/password for the testing cluster.

**Comment by lholmquist** [comment]
does this acutally work?  I tried to use this against my local minishift,  my first test was just using the default behavior to query the list of Projects, which came back succesfully.

The next test i did was to provide a username and password, but i got a permission error: `forbidden: User "system:anonymous" cannot`

**Comment by tremes** [comment]
Hmm I haven't tried it against minishift. I tried it against "real" cluster and it worked fine for me. I can try out minishift. I don't know whether minishift exposes exactly the same API TBH.

**Comment by tremes** [comment]
@lholmquist It works for me on minishift as well. I was using following config:
```
const customConfig = {
    apiVersion: 'v1',
    context:
    { cluster: process.env.CLUSTER_URL,
      namespace: '',
      user: 'admin' 
    },
    user: { 
      username: 'admin',
      password: 'admin' 
    },
    cluster: process.env.CLUSTER_URL 
  };
```
Can you share your test and config please?

**Comment by lholmquist** [comment]
looks like i wasn't using your changes in my test,  so i'll try again :)

File: lib/common-request.js [hunk 2]
```diff
@@ -28,14 +28,31 @@ const privates = require('./private-map');
 const request = require('request');
 
 module.exports = function doRequest (client, options) {
+  const clientConfig = privates.get(client).config;
+  if (clientConfig.user.token !== undefined) {
+    return doRequestWithToken(client, options, clientConfig.user.token);
+  } else {
+    return doRequestWithBasicAuth(client).then((accessToken) => {
```
Comment (lholmquist) [inline lib/common-request.js]: you also need to pass the options here.  i was getting a self-sign cert error from the request library even though i was passing in the "request options"

look at the doRequestWithToken function

Comment (tremes) [reply lib/common-request.js]: Oh yes I see. Thanks Luke. I'll update it.

Comment (tremes) [reply lib/common-request.js]: @lholmquist Hmm I am not sure TBH. Did you mean somthing like https://github.com/tremes/openshift-rest-client/commit/22b65f18e89cc45ff22f4e9b9f0390c73bc749bf ?

Comment (lholmquist) [reply lib/common-request.js]: @tremes very close,  you also need the requests settings to be merged,  something like this https://github.com/lholmquist/openshift-rest-client/commit/20bfbab219819892f291de29f86dd51233239938

i combined the headers and the base options in that commit,  but i like what you did separating them instead :)

Comment (tremes) [reply lib/common-request.js]: @lholmquist ahh ok thanks. So I updated it like this https://github.com/tremes/openshift-rest-client/commit/ca4f990efc9678b3dc0fc8e5db0e51046c9bb7b2. Will cherry-pick and rebase this PR if it's ok.

Comment (lholmquist) [reply lib/common-request.js]: no problem

Comment (tremes) [reply lib/common-request.js]: Done. Thanks. I am looking forward to a new release:-)

**Pull request merged**